miniapps/dpg/pconvection-diffusion.cpp

```diff
@@ -0,0 +1,788 @@
+
+         if (prob == prob_type::EJ)
+         {
+            ess_bdr_uhat = 0;
+            ess_bdr_fhat = 1;
+            ess_bdr_uhat[1] = 1;
+            ess_bdr_fhat[1] = 0;
+         }
+         else
+         {
+            ess_bdr_uhat = 1;
+            ess_bdr_fhat = 0;
+         }
+
+         hatu_fes->GetEssentialTrueDofs(ess_bdr_uhat, ess_tdof_list_uhat);
+         hatf_fes->GetEssentialTrueDofs(ess_bdr_fhat, ess_tdof_list_fhat);
+      }
+
+      // shift the ess_tdofs
+      int n = ess_tdof_list_uhat.Size();
+      int m = ess_tdof_list_fhat.Size();
+      Array<int> ess_tdof_list(n+m);
+      for (int j = 0; j < n; j++)
+      {
+         ess_tdof_list[j] = ess_tdof_list_uhat[j]
+                            + u_fes->GetTrueVSize()
+                            + sigma_fes->GetTrueVSize();
+      }
+      for (int j = 0; j < m; j++)
+      {
+         ess_tdof_list[j+n] = ess_tdof_list_fhat[j]
+                              + u_fes->GetTrueVSize()
+                              + sigma_fes->GetTrueVSize()
+                              + hatu_fes->GetTrueVSize();
+      }
+
+      Array<int> offsets(5);
+      offsets[0] = 0;
+      offsets[1] = u_fes->GetVSize();
+      offsets[2] = sigma_fes->GetVSize();
+      offsets[3] = hatu_fes->GetVSize();
+      offsets[4] = hatf_fes->GetVSize();
+      offsets.PartialSum();
+      BlockVector x(offsets);
+      x = 0.0;
+      hatu_gf.MakeRef(hatu_fes,x.GetBlock(2));
+      FunctionCoefficient bdr_cf(bdr_data);
+      hatu_gf.ProjectBdrCoefficient(bdr_cf,ess_bdr_uhat);
+
+      hatf_gf.MakeRef(hatf_fes,x.GetBlock(3));
+      hatf_gf.ProjectBdrCoefficientNormal(hatfex,ess_bdr_fhat);
+
+      OperatorPtr Ah;
+      Vector X,B;
+      a->FormLinearSystem(ess_tdof_list,x,Ah,X,B);
+
+      BlockOperator * A = Ah.As<BlockOperator>();
+
+
+      BlockDiagonalPreconditioner M(A->RowOffsets());
+      M.owns_blocks = 1;
+      int skip = 0;
+      if (!static_cond)
+      {
+         HypreBoomerAMG * amg0 = new HypreBoomerAMG((HypreParMatrix &)A->GetBlock(0,0));
+         HypreBoomerAMG * amg1 = new HypreBoomerAMG((HypreParMatrix &)A->GetBlock(1,1));
+         amg0->SetPrintLevel(0);
+         amg1->SetPrintLevel(0);
+         M.SetDiagonalBlock(0,amg0);
+         M.SetDiagonalBlock(1,amg1);
+         skip = 2;
+      }
+      HypreBoomerAMG * amg2 = new HypreBoomerAMG((HypreParMatrix &)A->GetBlock(skip,
+                                                                               skip));
+      amg2->SetPrintLevel(0);
+      M.SetDiagonalBlock(skip,amg2);
+
+      HypreSolver * prec;
+      if (dim == 2)
+      {
+         prec = new HypreAMS((HypreParMatrix &)A->GetBlock(skip+1,skip+1), hatf_fes);
+      }
+      else
+      {
+         prec = new HypreADS((HypreParMatrix &)A->GetBlock(skip+1,skip+1), hatf_fes);
+      }
+      M.SetDiagonalBlock(skip+1,prec);
+
+      CGSolver cg(MPI_COMM_WORLD);
+      cg.SetRelTol(1e-12);
+      cg.SetMaxIter(200000);
```

Comment:
Is there no preconditioner that can keep the iterations under 100,000?

Reply:
heheh, this was propably a leftover from some experiments of mine. Fixed